src/transformers/models/ibert/utils/quant_modules.py

```diff
@@ -0,0 +1,538 @@
+from torch.nn import Linear as _linear
+from torch.nn import Embedding as _Embedding
+from torch.nn import Module, Parameter
+from .quant_utils import *
+
+from ....utils import logging
+logger = logging.get_logger(__name__)
+
+class QuantEmbedding(Module):
+    """
+    Class to quantize given Embedding layer
+
+    Parameters:
+    weight_bit : int
+        Bitwidth for quantized weights.
+    momentum : float, default 0.95
+        Momentum for updating the activation quantization range.
+    quant_mode : bool, default False
+        The mode for quantization. True for quantization.
+    """
+    def __init__(self,
+                 weight_bit,
+                 momentum=0.95,
+                 quant_mode=False):
+        super(QuantEmbedding, self).__init__()
+
+        self.weight_bit = weight_bit
+        self.momentum = momentum
+        self.quant_mode = quant_mode
+        self.percentile_mode = False
+
+        if not self.quant_mode:
+            self.weight_function = None
+        else:
+            self.weight_function = SymmetricQuantFunction.apply
+                 
+    def set_param(self, embedding):
+        self.num_embeddings = embedding.num_embeddings
+        self.embedding_dim = embedding.embedding_dim
+        self.padding_idx = embedding.padding_idx
+        self.max_norm = embedding.max_norm
+        self.norm_type = embedding.norm_type
+        self.scale_grad_by_freq = embedding.scale_grad_by_freq
+        self.sparse = embedding.sparse
+        self.weight = embedding.weight
+
+        self.register_buffer('weight_scaling_factor', torch.zeros(1))
+        self.register_buffer('weight_integer', torch.zeros_like(self.weight))
+
+    def forward(self, x, positions=None, incremental_state=None):
+        if not self.quant_mode:
+            return F.embedding(
+                x,
+                self.weight,
+                self.padding_idx,
+                self.max_norm,
+                self.norm_type,
+                self.scale_grad_by_freq,
+                self.sparse,
+            ), None
+
+        w = self.weight
+        w_transform = w.data.detach()
+        w_min = w_transform.min().expand(1)
+        w_max = w_transform.max().expand(1)
+
+        self.weight_scaling_factor = symmetric_linear_quantization_params(
+                    self.weight_bit, w_min, w_max, False)
+        self.weight_integer = self.weight_function(
+                    self.weight, self.weight_bit, self.percentile_mode, 
+                    self.weight_scaling_factor)
+
+        emb_int = F.embedding(
+            x,
+            self.weight_integer,
+            self.padding_idx,
+            self.max_norm,
+            self.norm_type,
+            self.scale_grad_by_freq,
+            self.sparse,
+        )
+        return emb_int * self.weight_scaling_factor, self.weight_scaling_factor
+
+
+class QuantAct(Module):
+    """
+    Class to quantize given activations
+
+    Parameters:
+    ----------
```

Comment:
the `---------------------------` does not seem to correspond to our usual docstring. Could we maybe try to have it more in the format of: https://github.com/huggingface/transformers/blob/570218878a502469f2ce62a33bc62be9b3f89b4a/src/transformers/modeling_utils.py#L1251